src/bsp/src/stm32/src/InterlocManager.cpp

```diff
@@ -66,3 +82,174 @@ bool InterlocManager::constructUWBHeader(uint16_t destinationId,
+            // P type control, P  = 0.9
+            // TODO could had a PID type control #futureImprovement
+            dwCountOffset = 0.9 * error * DW_INTERNAL_CLOCK_RFEQ / SPEED_OF_LIGHT / 100;
+            // antenna delay is considered equal on Rx and Tx
+            device.setTxAntennaDLY((dwCountOffset >> 1) + device.getTxAntennaDLY());
+            device.setRxAntennaDLY((dwCountOffset >> 1) + device.getRxAntennaDLY());
+            if (dwCountOffset >> 1 <= 1) {
+                device.setState(DW_STATE::CALIBRATED);
+                m_calibFinishedCallback(m_calibFinishedCallbackContext);
+            }
+        }
+        Task::delay(50);
+    }
+}
+
+double InterlocManager::receiveTWRSequence(uint16_t destinationId, Decawave& device) {
+    UWBRxFrame rxFrame;
+    uint64_t pollRxTs;
+    uint64_t respTxTs;
+    uint64_t finalRxTs;
+    UWBMessages::TWRResponse responseMsg{};
+    double varA;
+    double varB;
+    double varC;
+    double varD;
+    double distanceEval;
+    double tof;
+    uint64_t tofDtu;
+
+    // receive poll message
+    device.receive(rxFrame, 0);
+
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_POLL) {
+        return 0;
+    }
+
+    pollRxTs = rxFrame.m_rxTimestamp;
+
+    // construct and send Response message
+    constructUWBHeader(destinationId, UWBMessages::DATA, UWBMessages::TWR_RESPONSE,
+                       (uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    respTxTs = (pollRxTs + (POLL_RX_TO_RESP_TX_DLY_UUS * UUS_TO_DWT_TIME))>>8;
+
+    device.transmit((uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    device.transmitDelayed((uint8_t*)&responseMsg,sizeof(responseMsg),respTxTime);
+
+    // wait for Final message reception
+    device.receive(rxFrame, 0);
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_FINAL) {
+        return 0;
+    }
+
+    UWBMessages::TWRFinal* finalTWRFrame = (UWBMessages::TWRFinal*)(rxFrame.m_frame);
+    device.getTxTimestamp(&respTxTs);
+    finalRxTs = rxFrame.m_rxTimestamp;
+
+    // evaluate distance
+    varA = (double)finalTWRFrame->m_respMinPoll;
+    varB = (double)((uint32_t)finalRxTs - (uint32_t)respTxTs);
+    varC = (double)(finalTWRFrame->m_finaleMinResp);
+    varD = (double)((uint32_t)respTxTs - (uint32_t)pollRxTs);
+
+    tofDtu = ((varA * varB - varC * varD) / (varA + varB + varC + varD));
+
+    tof = tofDtu * DWT_TIME_UNITS;
+    distanceEval = tof * SPEED_OF_LIGHT;
+
+    return distanceEval;
+}
+
+bool InterlocManager::sendTWRSequence(uint16_t destinationId, Decawave& device) {
+
+    UWBRxFrame responseFrame;
+    UWBRxFrame finalFrame;
+
+    UWBMessages::TWRPoll pollMsg{};
+    UWBMessages::TWRFinal finalMsg{};
+
+    uint64_t responseTimestamp;
+    uint64_t pollTimestamp;
+    uint64_t finalTxTime;
+    uint64_t finalTxTs;
```

Comment:
don't declare uninitialized variable unless you need to (eg. a while loop), initialize them at the same time you write on them for the first time.